lecture-notes/2-vcs.md

```diff
@@ -1859,10 +1859,126 @@ Running `git fetch` will update all of your remote tracking branches. For
+  (use "git pull" to update your local branch)
+$
+```
+
+As expected, the remote has updates that we do not, so our local branch is
+*behind*. The status message also notes that our local branch can be
+*fast-forwarded*, which we talked about earlier as well. This is the easiest
+case and is illustrated by the below diagram:
+
+```
+* a879406 (origin/mb-vcs-4) A shiny new thing, freshly downloaded
+* 5e72155 (mb-vcs-4) My most recent local commit
+* 2187cc4 My previous local commit
+.
+.
+.
+```
+
+It's not the *only* situation we could be in, though. We could also have a
+situation where the remote tracking branch has a new commit **and** the local
+branch has a *different* new commit. In that case, Git will note that the
+branches have "diverged" and will not let you push to the remote branch:
+
+```console?prompt=$
+$ g push
+To github.com:tekknolagi/isdt.git
+ ! [rejected]        mb-vcs-4 -> mb-vcs-4 (non-fast-forward)
+error: failed to push some refs to 'github.com:tekknolagi/isdt.git'
+hint: Updates were rejected because the tip of your current branch is behind
+hint: its remote counterpart. Integrate the remote changes (e.g.
+hint: 'git pull ...') before pushing again.
+hint: See the 'Note about fast-forwards' in 'git push --help' for details.
+$
+```
+
+Now, divergence only means that one ref is not strictly behind or in front of
+the other; it does not imply anything about the commit contents or whether the
+commits will textually conflict with one another. (For now, we'll assume *no
+conflicts*, because it's not the point of this section. The same principles
+from earlier still apply.)
+
+In either case, we can use rebase to bring in the remote changes. For example:
+
+```console?prompt=$
+$ git rebase origin/mb-vcs-4
+Successfully rebased and updated refs/heads/mb-vcs-4.
+$
+```
+
+This rebase command finds the *fork point* between the local `mb-vcs-4` and the
+remote `origin/mb-vcs-4`. Then, it applies (cherry-picks) all of the remote
+commits to the fork point. Last, it applies all of the local commits on top.
+
+If there are no local commits, Git will choose to automatically fast-forward
+the local branch to get it up to date with the remote branch. Otherwise, it
+will go through the full rigamarole of applying commits one by one (making new
+commits along the way!).
+
+After you have rebased, you can safely push your (renewed) local commits to the
+remote. That is, unless someone else has beaten you to the punch and pushed
+even more commits to the remote. In that case, keep fetching and rebasing.
+
+Note that this re-applying of local commits on top of remote changes
+constitutes *rewriting history*. Some people find this very distasteful and
+even ban it in their projects, preferring merge commits exclusively. The course
+staff has a more moderate view of things: rewriting history is fine if it's a
+private, short-lived development branch. Otherwise, merge.
```

Comment:
I don't think this is the right place to discuss history rewriting. It needs more than one paragraph to get sufficient nuance and honestly probably deserves an entire section. I've already written quite a bit on it in lecture 2, so I'd want the coverage here to build on what I say there instead of starting fresh.

Reply:
new section about incorporating main branch changes into feature branch and all associated problems

Reply:
are you doing this section in your section or should we put it in lecture 5 in the social practices part?

Reply:
Added TODO